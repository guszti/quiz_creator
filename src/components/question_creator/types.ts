export type QuestionCreatorState = {
    title: string;
    answers: Answer[];
};

export type Question = {
    title: string;
    answers: Answer[];
};

export type Answer = {
    text: string;
    isRight: boolean;
};
