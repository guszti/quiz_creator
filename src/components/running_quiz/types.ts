export type QuizState = {
    result: number;
    score: number;
    questionNumber: number;
};
