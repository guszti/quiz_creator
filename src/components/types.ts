import { Question } from "./question_creator/types";

export type GlobalState = {
    isRunning: boolean;
    questions: Question[];
};
