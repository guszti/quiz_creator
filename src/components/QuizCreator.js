import React from 'react';
import Quiz from './Quiz';
import QuestionForm from './QuestionForm';
import Answer from './Answer';
import TitleBar from './TitleBar'

class QuizCreator extends React.Component{
    constructor(props){
      super(props);
  
      this.state = {
        title: '',
        answers: [],
        questions: [],
        running: false
      }
  
      this.add_title = this.add_title.bind(this);
      this.add_answer = this.add_answer.bind(this);
      this.save_question = this.save_question.bind(this);
      this.mark_answer = this.mark_answer.bind(this);
      this.run_quiz = this.run_quiz.bind(this);
      this.reset_quiz = this.reset_quiz.bind(this);
    }
  
    add_title(title){
      this.setState({
        title: title
      });
    }
  
    add_answer(answer){
      const new_answer = {
        text: answer,
        right: false
      };
  
      this.setState({
        answers: this.state.answers.concat([new_answer])
      });
    }
  
    save_question(){
      let ans = this.state.answers.filter((item) => item.right === true);
  
      if(ans.length > 0){
        this.setState({
          questions: this.state.questions.concat([{'title': this.state.title, 'answers': this.state.answers}]),
          title: '',
          answers: []
        }, () => {console.log(this.state.questions);});
      }else{
        alert('Mark at least one good answer!');
      }
    }
  
    mark_answer(m, i){
      let new_list = this.state.answers;
      new_list[i].right = m;
  
      this.setState({
        answers: new_list
      });
      console.log(this.state.answers);
    }
  
    run_quiz(){
      this.setState({
        running: true
      });
    }

    reset_quiz(){
      this.setState({
        running: false,
        questions: []
      });
    }

    render(){
  
      const save = (
          <button onClick={this.save_question}>save question</button>
      );
  
      const run = (
        <button onClick={this.run_quiz}>begin quiz</button>
      );
  
      return(
        <div>
          <TitleBar />
          {this.state.running ? <Quiz questions={this.state.questions} reset={this.reset_quiz} /> :
          <div class="container">
            {this.state.questions.length >= 1 ? run : ''}
            <h1>{this.state.title}</h1>
            <QuestionForm t={this.state.title} add_title={this.add_title} add_answer={this.add_answer} />
            <ul>
              {this.state.answers.map((a, i) =>
                <Answer a={a.text} key={i} index={i} mark_answer={this.mark_answer} />
              )}
            </ul>
            {this.state.answers.length >= 2 ? save : ''}
            {this.state.questions.map((item, i) => 
            <div key={i}>
              <br />
              <br/>
              <hr />
              <table>
                <tbody>
                  <tr><th>{item.title}</th></tr>
                  {item.answers.map((a, j) => 
                    <tr key={j}>
                      {(a.right === true) ? <td style={{backgroundColor:'Chartreuse'}}><b>{a.text}</b></td> : <td>{a.text}</td>}
                    </tr>
                  )}
                </tbody>
              </table>
              </div>
            )}
          </div>}
        </div>
      );
    }
  }

  export default QuizCreator;