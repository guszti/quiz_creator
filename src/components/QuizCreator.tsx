import * as React from "react";
import { FC } from "react";
import TitleBar from "./common/TitleBar";
import Quiz from "./running_quiz/Quiz";
import { useSelector } from "react-redux";
import { getIsRunning } from "./redux/selectors";
import QuestionCreator from "./question_creator/QuestionCreator";

const QuizCreator: FC = () => {
    const isRunning = useSelector(getIsRunning);

    return (
        <div>
            <TitleBar />
            {isRunning ? <Quiz /> : <QuestionCreator />}
        </div>
    );
};

export default QuizCreator;
