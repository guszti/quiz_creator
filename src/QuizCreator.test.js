import React from 'react';
import ReactDOM from 'react-dom';
import QuizCreator from './App';
import { create, update } from 'react-test-renderer'
import { exportAllDeclaration } from '@babel/types';

describe('Starting the quiz', () =>{
    test('toggle button', () => {
  
      let tree = create(<QuizCreator />);
  
      let instance = tree.getInstance();
  
      expect(instance.state.running).toBe(false);
  
      instance.run_quiz();
  
      expect(instance.state.running).toBe(true);
  
      expect(tree.toJSON()).toMatchSnapshot();
    
    });
  });

  describe('Saving a question', () => {
      test('click save', () => {

        let tree = create(<QuizCreator />);

        let instance = tree.getInstance();

        instance.add_title('asd');

        instance.add_answer('asd');

        instance.add_answer('asd');

        instance.mark_answer(true, 0);

        instance.save_question();

        expect(instance.state.questions.length).toBe(1);

      });
  });