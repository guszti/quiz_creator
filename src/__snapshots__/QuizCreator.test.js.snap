// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Starting the quiz toggle button 1`] = `
<div>
  <div>
    <h2>
      Quiz has ended!
    </h2>
    <p>
      You've reached 
      0
      %
    </p>
    <br />
    <button>
      Create new quiz
    </button>
    <button>
      Try this one again
    </button>
  </div>
</div>
`;
